designs/v1-api.md

```diff
@@ -0,0 +1,529 @@
+2. NodePool Name
+3. NodeClass Name
+
+### Changes to the API
+
+#### NodeClaim Spec Immutability
+
+**Category:** Breaking
+
+Karpenter currently doesn’t enforce immutability on NodeClaims in v1beta1, though we implicitly assume that users should not be acting against these objects after creation, as the NodeClaim lifecycle controller won’t react to any change after the initial instance launch.
+
+Karpenter can make every `spec` field immutable on the NodeClaim after its initial creation. This will be enforced through CEL validation, where you can perform a check like `[self == oldSelf](https://kubernetes.io/docs/reference/using-api/cel/#language-overview)` to enforce that the fields cannot have changed after the initial apply. Users who are not on K8s 1.25+ that supports CEL will get the same validation enforced by validating webhooks.
+
+## EC2NodeClass API [AWS CloudProvider]
+
+```
+apiVersion: karpenter.k8s.aws/v1
+kind: EC2NodeClass
+metadata:
+  name: default
+spec:
+  kubelet:
+    podsPerCore: 2
+    maxPods: 20
+    systemReserved:
+        cpu: 100m
+        memory: 100Mi
+        ephemeral-storage: 1Gi
+    kubeReserved:
+        cpu: 200m
+        memory: 100Mi
+        ephemeral-storage: 3Gi
+    evictionHard:
+        memory.available: 5%
+        nodefs.available: 10%
+        nodefs.inodesFree: 10%
+    evictionSoft:
+        memory.available: 500Mi
+        nodefs.available: 15%
+        nodefs.inodesFree: 15%
+   evictionSoftGracePeriod:
+       memory.available: 1m
+       nodefs.available: 1m30s
+       nodefs.inodesFree: 2m
+   evictionMaxPodGracePeriod: 60
+   imageGCHighThresholdPercent: 85
+   imageGCLowThresholdPercent: 80
+   cpuCFSQuota: true
+   clusterDNS: ["10.0.1.100"]
+  bootstrapMode: Bottlerocket
+  subnetSelectorTerms:
+    - tags:
+        karpenter.sh/discovery: "${CLUSTER_NAME}"
+    - id: subnet-09fa4a0a8f233a921
+  securityGroupSelectorTerms:
+    - tags:
+        karpenter.sh/discovery: "${CLUSTER_NAME}" 
+    - name: my-security-group
+    - id: sg-063d7acfb4b06c82c
+  amiSelectorTerms:
+    - tags:
+        karpenter.sh/discovery: "${CLUSTER_NAME}"
+    - name: my-ami
+    - id: ami-123
+  role: "KarpenterNodeRole-${CLUSTER_NAME}"
+  instanceProfile: "KarpenterNodeInstanceProfile-${CLUSTER_NAME}"
+  userData: |
+    echo "Hello world"
+  tags:
+    team: team-a
+    app: team-a-app
+  instanceStorePolicy: RAID0
+  metadataOptions:
+    httpEndpoint: enabled
+    httpProtocolIPv6: disabled
+    httpPutResponseHopLimit: 1 # This is changed to disable IMDS access from containers not on the host network
+    httpTokens: required
+  blockDeviceMappings:
+    - deviceName: /dev/xvda
+      ebs:
+        volumeSize: 100Gi
+        volumeType: gp3
+        iops: 10000
+        encrypted: true
+        kmsKeyID: "1234abcd-12ab-34cd-56ef-1234567890ab"
+        deleteOnTermination: true
+        throughput: 125
+        snapshotID: snap-0123456789
+  detailedMonitoring: **true**
+status:
```

Comment:
Doesn't block v1, but: could we add the ARN of the instance profile into `.status`?

Reply:
What's the need for the ARN vs. just the name? The instance profile will be in the same partition and is global, so I think the ARN should be easily constructed from the name